### DISCUSSION TITLE
How to color the popup border and separators of a menu?

### DISCUSSION THREAD

**Discussion by Pyrdacor** [AvaloniaUI/Avalonia#7572] [Q&A]
I managed to customize my menu like this:

![image](https://user-images.githubusercontent.com/8607784/153266196-cdbd2477-4778-4f3d-a967-c584a7186962.png)

But as you can see there is still this white border around the popup and the separator has white background as well. How can I color those areas?

This is my styling:

```xaml
<Style Selector="MenuItem">
	<Setter Property="Foreground" Value="#aaaa99"/>
	<Setter Property="Background" Value="#666655"/>
</Style>
<Style Selector="MenuItem /template/ ContentPresenter#PART_HeaderPresenter">
	<Setter Property="ContentTemplate">
		<DataTemplate>
			<views:Label Text="{Binding, Converter={StaticResource MenuTextConverter}}"/>
		</DataTemplate>
	</Setter>
</Style>
<Style Selector="MenuItem:selected /template/ ContentPresenter#PART_HeaderPresenter">
	<Setter Property="TextBlock.Foreground" Value="#ffffee" />
</Style>
<Style Selector="MenuItem:open /template/ ContentPresenter#PART_HeaderPresenter">
	<Setter Property="TextBlock.Foreground" Value="#ffffee" />
</Style>
<Style Selector="MenuItem > MenuItem:pointerover /template/ ContentPresenter#PART_HeaderPresenter">
	<Setter Property="TextBlock.Background" Value="#888877" />
</Style>
<Style Selector="MenuItem > MenuItem:pointerover /template/ Border#PART_LayoutRoot">
	<Setter Property="Background" Value="#888877" />
</Style>
<Style Selector="MenuItem > MenuItem /template/ Border#PART_LayoutRoot">
	<Setter Property="BorderBrush" Value="#222222" />
</Style>
<Style Selector="MenuItem:separator">
	<Setter Property="Background" Value="#666655" />
</Style>
<Style Selector="MenuItem > Separator">
	<Setter Property="Background" Value="#666655" />
</Style>
<Style Selector="MenuItem /template/ Popup#PART_Popup">
	<Setter Property="Border.Background" Value="#666655" />
</Style>
```

The last 4 styles were my attempt to change the areas. But none of them worked.

The menu looks like this:

```xaml
<Border DockPanel.Dock="Top" 
		BorderBrush="#222222"
		BorderThickness="1"
		BoxShadow="3 3 Black">
	<Menu Background="#666655" BorderBrush="#ff0000">
		<MenuItem Header="_File">
			<MenuItem Header="_Create..." Command="{Binding CreateCommand}"/>
			<MenuItem Header="_Open..." Command="{Binding OpenCommand}"/>
			<MenuItem Header="_Save..." Command="{Binding SaveCommand}"/>
			<Separator/>
			<MenuItem Header="_Exit" Command="{Binding QuitCommand}"/>
		</MenuItem>
		<MenuItem Header="_Edit">
			<MenuItem Header="Copy"/>
			<MenuItem Header="Paste"/>
		</MenuItem>			
	</Menu>
</Border>
```

FYI: `views:Label` is just a TextBlock with a shadow (transformed second TextBlock). And `MenuTextConverter` just removes the underscores for display.

**Comment by Pyrdacor** [comment]
Might be related to #7571.

**Reply by pagner** [reply]
@Pyrdacor Did you ever manage to resolve this? I have the same issue with the separator in a menu when a margin is applied. There is a different background color in the margin around the separator.

**Comment by Pyrdacor** [comment]
I don't think so unfortunately. :(

**Comment by timunie** [comment]
Hi @Pyrdacor 

I may have a solution for you. The issue is, that the Background has a local value set which cannot be overriden by any Style. This may change in the future, but for now this is not possible without recreating the `ControlTemplate`

This is the line which is bugging you: https://github.com/AvaloniaUI/Avalonia/blob/12ecfde832b22fea97f11369c784f20075fb66d7/src/Avalonia.Themes.Fluent/Controls/MenuItem.xaml#L118

The good news is, that it is a `DynamicResource`. So we can override it just like this: 

```xml
<!-- Can be any Resources section -->
<Window.Resources>
  <SolidColorBrush x:Key="MenuFlyoutPresenterBackground">Red</SolidColorBrush>
</Window.Resources>
```
 
The result is: 
![image](https://user-images.githubusercontent.com/47110241/168766972-f6330cd1-43b4-4ebc-b56b-0067da8d4a04.png)


Happy coding
Tim 

/cc @pagner

**Reply by pagner** [reply]
@timunie Thanks for the reply. Unfortunately this did not work for me. The Border background has no effect at all for my case. 

I tried your way using Resources. I also tried using styles. You can drill down into the Border using this:

`<Style Selector="MenuItem /template/ Popup#PART_Popup > Border">`

I can change the padding and other properties. Background has no effect.

**Reply by timunie** [reply]
My sample is for Fluent Theme. If you have another, you need to lookup in the source

**Reply by pagner** [reply]
@timunie thanks for pointing that out. I didn't notice what theme you are using. I am using the Default Theme which unfortunately uses ThemeBackgroundBrush

**Reply by timunie** [reply]
You can still override this. Just make sure it's on the specific styles resources overridden iirc

**Comment by Pyrdacor** [comment]
Thanks. Will have a look when I am on to that project again. ;)

**Comment by pagner** [comment]
@Pyrdacor I was able to override the style template and fix my issue.

https://github.com/AvaloniaUI/Avalonia/blob/12ecfde832b22fea97f11369c784f20075fb66d7/src/Avalonia.Themes.Default/Controls/MenuItem.xaml#L93

Thanks again @timunie for your help

**Comment by Rabbex1** [comment]
@Pyrdacor I was able to achive what you were looking for with 

```xml
<Window.Styles>
  <Style Selector="MenuItem:open /template/ Popup#PART_Popup > Border">
	  <Setter Property="BorderBrush" Value="red" />
	  <Setter Property="BorderThickness" Value="1" />
	  <Setter Property="CornerRadius" Value="0" />
	  <Setter Property="Background" Value="blue"></Setter>
  </Style>
</Window.Styles>

```
